components/service/glean/src/test/java/mozilla/components/service/glean/storages/CustomDistributionsStorageEngineTest.kt

```diff
@@ -0,0 +1,512 @@
+
+    @get:Rule
+    val gleanRule = GleanTestRule(ApplicationProvider.getApplicationContext())
+
+    @Test
+    fun `accumulate() properly updates the values in all stores`() {
+        val storeNames = listOf("store1", "store2")
+
+        val metric = CustomDistributionMetricType(
+            disabled = false,
+            category = "telemetry",
+            lifetime = Lifetime.Ping,
+            name = "test_custom_distribution",
+            sendInPings = storeNames,
+            range = listOf(0L, 60000L),
+            bucketCount = 100,
+            histogramType = HistogramType.Exponential
+        )
+
+        // Create a sample that will fall into the underflow bucket (bucket '0') so we can easily
+        // find it
+        val sample = 1L
+        CustomDistributionsStorageEngine.accumulateSamples(
+            metricData = metric,
+            samples = listOf(sample).toLongArray(),
+            rangeMin = 0L,
+            rangeMax = 60000L,
+            bucketCount = 100,
+            histogramType = HistogramType.Exponential
+        )
+
+        // Check that the data was correctly set in each store.
+        for (storeName in storeNames) {
+            val snapshot = CustomDistributionsStorageEngine.getSnapshot(
+                storeName = storeName,
+                clearStore = true
+            )
+            assertEquals(1, snapshot!!.size)
+            assertEquals(1L, snapshot["telemetry.test_custom_distribution"]?.values!![1])
+        }
+    }
+
+    @Test
+    fun `deserializer should correctly parse custom distributions`() {
+        // We are using the CustomDistributionData here as a way to turn the object
+        // into JSON for easy comparison
+        val metric = CustomDistributionMetricType(
+            disabled = false,
+            category = "telemetry",
+            lifetime = Lifetime.Ping,
+            name = "test_custom_distribution",
+            sendInPings = listOf("store1", "store2"),
+            range = listOf(0L, 60000L),
+            bucketCount = 100,
+            histogramType = HistogramType.Exponential
+        )
+        val td = CustomDistributionData(
+            category = metric.category,
+            name = metric.name,
+            rangeMin = 0L,
+            rangeMax = 60000L,
+            bucketCount = 100,
+            histogramType = HistogramType.Exponential
+        )
+
+        val persistedSample = mapOf(
+            "store1#telemetry.invalid_string" to "invalid_string",
+            "store1#telemetry.invalid_bool" to false,
+            "store1#telemetry.null" to null,
+            "store1#telemetry.invalid_int" to -1,
+            "store1#telemetry.invalid_list" to listOf("1", "2", "3"),
+            "store1#telemetry.invalid_int_list" to "[1,2,3]",
+            "store1#telemetry.invalid_td_name" to "{\"category\":\"telemetry\",\"bucket_count\":100,\"range\":[0,60000,12],\"histogram_type\":1,\"values\":{},\"sum\":0,\"time_unit\":2}",
```

Comment:
To make this test reflect the implementation, we need to remove the `time_unit` entry from each of the persisted samples here and below.